modules/javafx.graphics/src/main/native-glass/gtk/glass_key.cpp

```diff
@@ -345,4 +345,36 @@ JNIEXPORT jint JNICALL Java_com_sun_glass_ui_gtk_GtkApplication__1getKeyCodeForC
     return gdk_keyval_to_glass(keyval);
 }
 
+/*
+ * Class:     com_sun_glass_ui_gtk_GtkApplication
+ * Method:    _isKeyLocked
+ * Signature: (I)I
+ */
+JNIEXPORT jint JNICALL Java_com_sun_glass_ui_gtk_GtkApplication__1isKeyLocked
+  (JNIEnv * env, jobject obj, jint keyCode)
+{
+#ifdef GLASS_GTK3
+    GdkKeymap *keyMap = gdk_keymap_get_default();
+    gboolean lockState = FALSE;
+    switch (keyCode) {
+        case com_sun_glass_events_KeyEvent_VK_CAPS_LOCK:
+            lockState = gdk_keymap_get_caps_lock_state(keyMap);
```

Comment:
It seems that `gdk_keymap_get_caps_lock_state` isn't reliable on GTK 3 either. At least not when I test it on a Linux VM. I may end up using raw Xlib calls instead (which would have the added benefit of removing the "ifdef" for GTK 3).

Reply:
I did end going this route. Instead of the `gdk_keymap` functions, I now use `XkbGetNamedIndicator`. It works on both GTK2 and GTK3 (unsurprisingly, since it isn't using GTK / GDK at all), and correctly reports the state of the keys even if the lock state is toggled when the program is not running or does not have focus.